lib/common/abstract-descriptor.h

```diff
@@ -0,0 +1,269 @@
+  // into subscripts.
+  void GetLowerBounds(SubscriptArray &subscripts) const {
+    DESC::GetLowerBounds(subscripts);
+  }
+
+  bool HasSameTypeAs(const DescriptorInterface<DESC> &other) const {
+    return DESC::HasSameTypeAs(static_cast<const DESC &>(other));
+  }
+
+  // When the decsriptor describes an array containing integer,
+  // GetSubscriptValueAt extract the element indexed by subscripts and convert
+  // it to a SubscriptValue.
+  SubscriptValue GetSubscriptValueAt(const SubscriptArray &subscripts) const {
+    return DESC::GetSubscriptValueAt(subscripts);
+  }
+
+  // Number of elements
+  SizeValue Elements() const { return DESC::Elements(); }
+
+  // Create a descriptor with the same type (including type parameters if
+  // applicable) as source and the given rank. The data and bounds are left
+  // unset.
+  static OwningPointer<DescriptorInterface<DESC>> CreateWithSameTypeAs(
+      const DescriptorInterface<DESC> &source, Rank rank, Attribute attribute) {
+    return OwningPointer<DescriptorInterface<DESC>>{
+        static_cast<DescriptorInterface<DESC> *>(&(*DESC::CreateWithSameTypeAs(
+            static_cast<const DESC &>(source), rank, attribute)))};
+  }
+
+  // Allocate a descriptor for which the type is fully defined:
+  //  - set bounds information (the lower bounds are set to to 1)
+  //  - get storage for the entity described by the descriptor and
+  //    make it available to the descriptor.
+  //  The data is not initialized (even for derived type).
+  void Allocate(const SubscriptArray &extents) { DESC::Allocate(extents); }
+
+  // Copy count elements from the entity described by source into the entity
+  // describe by the descriptor.
+  // - source and the descriptor must describe entities of the same type. source
+  // must have at least one element.
+  // - For the target, the copy starts at the element indexed by resultSubscript
+  // and resultSubscript are incremented according to the dimension order in
+  // dimOrder.
+  // - For the source, the copy starts at the first element and source elements
+  // are run through in Fortran dimension order. If the end of the source of the
+  // target is reached before count, the copy continues from the first element
+  // of the entity whose last element was reached.
+  SizeValue CopyFrom(const DescriptorInterface<DESC> &source,
+      SubscriptArray &resultSubscripts, SizeValue count,
+      const RankedSizedArray<Rank> *dimOrder) {
+    return DESC::CopyFrom(
+        static_cast<const DESC &>(source), resultSubscripts, count, dimOrder);
+  }
+
+  // Indicate that the descriptor is not describing a user variable (this
+  // information may not be useful and this may be a no-op)
+  void SetCompilerCreatedFlag() { DESC::SetCompilerCreatedFlag(); }
+
+  // TODO: improve interface for attributes
+  static Attribute AllocatableAttribute() {
+    return DESC::AllocatableAttribute();
+  }
+};
+
+namespace safestd {
```

Comment:
This seems likely to be useful in other situations and should be in its own header.